src/main/java/jenkins/plugins/logstash/configuration/LogstashIndexer.java

```diff
@@ -0,0 +1,156 @@
+package jenkins.plugins.logstash.configuration;
+
+import javax.annotation.Nonnull;
+
+import org.apache.commons.lang.StringUtils;
+import org.kohsuke.stapler.DataBoundSetter;
+import org.kohsuke.stapler.QueryParameter;
+import org.kohsuke.stapler.StaplerRequest;
+
+import hudson.DescriptorExtensionList;
+import hudson.ExtensionPoint;
+import hudson.model.AbstractDescribableImpl;
+import hudson.model.Descriptor;
+import hudson.model.ReconfigurableDescribable;
+import hudson.model.Descriptor.FormException;
+import hudson.util.FormValidation;
+import jenkins.model.Jenkins;
+import jenkins.plugins.logstash.Messages;
+import jenkins.plugins.logstash.persistence.AbstractLogstashIndexerDao;
+import net.sf.json.JSONObject;
+
+public abstract class LogstashIndexer<T extends AbstractLogstashIndexerDao>
+    extends AbstractDescribableImpl<LogstashIndexer<?>>
+    implements ExtensionPoint, ReconfigurableDescribable<LogstashIndexer<?>>
+{
+  protected String host;
```

Comment:
I'm not sure, but maybe it would be better to extract those fields into a separate data object?
It seems cleaner to push the check in `shouldRefreshInstance()` into the `equals()` method instead of overriding   `shouldRefreshInstance()` directly.

Reply:
But we're not comparing the same objects in this method so the `equals()` method is not the right place. Currently there is a separation between the configuration and the actual dao that does the push to the indexer. The configuration class must decide if it should create return a new dao instance or return the existing one which is done in the shouldRefreshInstance().

Reply:
What I don't like here is that LogstashIndexer is an abstract factory for the dao while also being the model for the configuration view (jelly). My intuition suggest extracting the model into a separate entity. 

Maybe let me send you a patch to illustrate (as it might be easier than explaining) and you can revert it if you don't like it.

Reply:
@mwinter69 can you check ae96695471bcf9ab8eeabddd24dc14e642f81947 ?
It's only a rough draft but I think it demonstrates the idea (e.g. I'd rather have the data object immutable than cloning it).

Please excuse me for the messed up formatting, I accidentally formatted the whole Syslog.java

Reply:
I think this makes the whole thing with stapler and configuration more complex. Not sure if we have to make the Data object then also a describable so that stapler can auto configure it when global configuration is saved.
We could also move the comparison to the GlobalConfiguration class.
A third option would be to make the daos directly the thing that is configured.

Reply:
With your explanation I think it's OK to leave the code as is. It's still better than previous one.

> We could also move the comparison to the GlobalConfiguration class.

Can you explain what you mean by this?

Reply:
Take a look at https://github.com/mwinter69/logstash-plugin/tree/configuration2
One effect of stapler is that when you bind a request to the configuration you always get a new instance of the logstashindexer. Thats why I had this check and implemented the reconfigure.
In this branch I replaced this by comparing (with equals) the newly configured instance with the active instance. If they are different, the active is replaced with the new one.

Reply:
I like this more since we are comparing instances of `LogstashIndexer` instead of comparing `LogstashIndexer` and a `LogstashDao`.